File: EAData/EADataCollector.py
#! /usr/bin/python3

import syslog
import datetime
import json
import boto3
import psycopg2
from urllib import request



# we are useing AWS Secrets Manager to hold API keys, and mqtt account password This retrieves a secret


def get_secret(MySecretString):

    secret_name = "RdgHydroServerSecrets"
    region_name = "eu-west-2"

    session = boto3.session.Session()
    client = session.client(service_name='secretsmanager', region_name=region_name)

    get_secret_value_response = client.get_secret_value(SecretId=secret_name)

    all_secret = json.loads(get_secret_value_response['SecretString'])
    secret = all_secret.get(MySecretString)
    return secret




def main():

    syslog.syslog("Starting update of EA data tables")
    print("Starting update of EA data tables")

    DbHost = get_secret("DB_HOST")
    DbPort = int(get_secret("DB_PORT"))
    DbUser = get_secret("DB_USER")
    DbPassword = get_secret("DB_PASSWORD")
    DbName = get_secret("DB_NAME")

    try:
        db = psycopg2.connect(dbname=DbName, user=DbUser, password=DbPassword, host=DbHost, port=DbPort)
    except Exception as e:
        print("Database error on connection: ",e)
        syslog.syslog("Database connection error",e)

# get the datetime of the last database entry
    
    cursor = db.cursor()
    cursor.execute("SELECT entrytime FROM public.eadata ORDER BY entrytime DESC")
    LastEntry = cursor.fetchone()[0]
    syslog.syslog("Last data entry is currently: " + LastEntry.strftime("%Y=%m-%d %H:%M:%S"))
    print("Last data entry is currently: ", LastEntry)

# get the flow data from the Enviroment Agency since last entry
    myds3 = LastEntry.strftime('%Y-%m-%d')
    myds4 = datetime.datetime.utcnow().strftime('%Y-%m-%d')
    EAapiURL = "https://environment.data.gov.uk/flood-monitoring/id/measures/2200TH-flow--Mean-15_min-m3_s/readings?startdate=" + myds3 + "&enddate=" + myds4 + "&_sorted&_limit=3000"
    try: 
        jsondata = request.urlopen(url=EAapiURL,timeout=20)
    except:
        print("Error getting EA data")
        syslog.syslog("Error getting EA data")
# enter the data into the database only if is after the LsatEntry (new data)

    rowsadded = 0
    data = json.loads(jsondata.read())
    items = data.get('items')
    for row in items:
        rowtime = datetime.datetime.strptime(row.get('dateTime'),"%Y-%m-%dT%H:%M:%SZ")
        rowtime = rowtime.replace(tzinfo=datetime.timezone.utc)
        if  rowtime > LastEntry:
            cursor.execute("INSERT INTO public.eadata (entrytime, flow) VALUES (%s, %s)", (row.get('dateTime'), row.get('value')))
            rowsadded += 1

# commit and close the database

    syslog.syslog("Rows added are: " + str(rowsadded))
    print("Rows added are: ",rowsadded)    
    db.commit()
    cursor.close()
    db.close()

if __name__ == '__main__':
    main()

File: EAData/EADataCollector.go
package main

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"ioutil"
	"log"
	"net/http"
	"strconv"
	"time"

	aws "github.com/MarkStanley/awssecrets"
	_ "github.com/lib/pq"
)

// PLCData - partial struct for a Hydro Message
type EAData struct {
	entrytime time.Time `json:"entrytime"`
	flow      float32   `json:"flow"`
}

func main() {
	aws.GetAllSecrets(SecretsName)

	ServerPort = aws.GetSecret("HYDRO_API_SERVER_PORT")

	DbHost := aws.GetSecret("DB_HOST")
	DbPort, _ := strconv.ParseInt(aws.GetSecret("DB_PORT"), 10, 0)
	DbUser := aws.GetSecret("DB_USER")
	DbPassword := aws.GetSecret("DB_PASSWORD")
	DbName := aws.GetSecret("DB_NAME")
	CertDir := aws.GetSecret("CERT_DIR")

	// Connect to the database
	connString := fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s sslmode=disable", DbHost, DbPort, DbUser, DbPassword, DbName)
	//log.Println(connString)
	db, err := sql.Open("postgres", connString)
	if err != nil {
		panic(err)
	}
	defer db.Close()

	err = db.Ping()
	if err != nil {
		panic(err)
	}

	//database form is:
	//    CREATE TABLE public.eadata (time timestamp NOT NULL, flow double precision);
	//	  ALTER TABLE public.eadata OWNER TO hydrodbuser;
	//	  ALTER TABLE ONLY public.eadata ADD CONSTRAINT eadata_pkey PRIMARY KEY (time);

	// Read the database and find the time of the last data value
	var Last EAData
	err = db.QueryRow("SELECT time FROM public.eadata ORDER BY entrytime DESC").Scan(&Last.entrytime, &Last.flow)
	if err != nil {
		log.Println(fmt.Errorf("Error retrieving current EA data: %v", err))
		return
	}

	// Read the EA data for the data items since the last data value

	// Fetch data from the API
	myds3 := Last.entrytime.Format("2001-01-02")
	myds4 := time.Now().Format("2001-01-02")
	EAapiURL := "https://environment.data.gov.uk/flood-monitoring/id/measures/2200TH-flow--Mean-15_min-m3_s/readings?startdate=" + myds3 + "&enddate=" + myds4 + "&_sorted&_limit=3000"
	response, err := http.Get(EAapiURL)
	if err != nil {
		log.Fatalf("Error fetching data from API: %v", err)
	}
	defer response.Body.Close()

	if response.StatusCode != http.StatusOK {
		log.Fatalf("API returned non-OK status: %s", response.Status)
	}

	// Read response body
	bodyBytes, err := ioutil.ReadAll(response.Body)
	if err != nil {
		log.Fatalf("Error reading response body: %v", err)
	}

	// Parse JSON data
	var items []Item
	err = json.Unmarshal(bodyBytes, &items)
	if err != nil {
		log.Fatalf("Error parsing JSON data: %v", err)
	}

	// Enter any new data into the database

}
